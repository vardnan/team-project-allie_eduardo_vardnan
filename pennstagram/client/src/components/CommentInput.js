import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import io from 'socket.io-client';
import submitComments from '../api/submitComment';

function CommentInput({
  postId,
  currentUser,
  // comments,
  // setComments,
}) {
  const [comment, setComment] = useState('');
  const textAreaRef = useRef(null);
  // const socket = io('http://localhost:8080');
  const socket = io('');

  const onSubmit = async () => {
    if (comment.trim() !== '') {
      const newCommentData = {
        username: currentUser.username,
        comment,
        commenterProfileImage: currentUser.profileImage,
        id: Math.random(),
      };

      try {
        await submitComments(postId, newCommentData);
        // Emit a new comment event to the backend
        socket.emit('newComment', newCommentData);
      } catch (err) {
        throw new Error(err);
      }

      // setComments([...comments, newCommentData]);
      setComment('');
    }
  };

  const keyDownHandler = (event) => {
    if (event.key === 'Enter' && !event.shiftKey) {
      event.preventDefault();
      onSubmit();
    }
  };

  return (
    <div>
      <textarea
        value={comment}
        ref={textAreaRef}
        onChange={(e) => setComment(e.target.value)}
        type="text"
        placeholder="Add a comment"
        onKeyDown={keyDownHandler}
      />
    </div>
  );
}

CommentInput.propTypes = {
  postId: PropTypes.isRequired,
  currentUser: PropTypes.object.isRequired,
  // comments: PropTypes.arrayOf(
  //   PropTypes.shape({
  //     username: PropTypes.string.isRequired,
  //     comment: PropTypes.string.isRequired,
  //   }),
  // ).isRequired,
  setComments: PropTypes.func.isRequired,
};

export default CommentInput;
