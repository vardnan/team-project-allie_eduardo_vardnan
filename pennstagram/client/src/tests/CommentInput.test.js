import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CommentInput from '../components/CommentInput';

// Testing rendering

test('Testing that the CommentInput is rendering correctly', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };

  const { getByRole } = render(<CommentInput {...props} />);
  expect(getByRole('textbox')).toBeInTheDocument();
});

test('Testing that the text area has the correct placeholder text', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  expect(getByPlaceholderText('Add a comment')).toBeInTheDocument();
});

test('Testing that text area is focused if focusOnLoad === true', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: true,
    setComments: () => {},
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);

  const textArea = getByPlaceholderText('Add a comment');

  expect(document.activeElement).toBe(textArea);
});

test('Testing that text area is unfocused if focusOnLoad === false', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);

  const textArea = getByPlaceholderText('Add a comment');

  expect(document.activeElement).not.toBe(textArea);
});

test('Testing that the text area has an initial empty value', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };
  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  expect(getByPlaceholderText('Add a comment')).toHaveValue('');
});

// Snapshot testing

test('Testing that the CommentInput matches snapshot', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };

  const component = renderer.create(<CommentInput {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// Testing events

test('Testing that the value of the comment is being updated as users are typing', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    focusOnLoad: false,
    setComments: () => {},
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  const textArea = getByPlaceholderText('Add a comment');

  fireEvent.change(textArea, { target: { value: 'New comment' } });

  expect(textArea.value).toBe('New comment');
});

test('Testing the triggering of onSubmit() when pressing "Enter" without also pressing the "Shift" key', () => {
  const mockSetComments = jest.fn();
  const props = {
    currentUser: 'Alice',
    comments: [],
    setComments: mockSetComments,
    focusOnLoad: false,
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  const textarea = getByPlaceholderText('Add a comment');

  fireEvent.change(textarea, { target: { value: 'New comment' } });
  fireEvent.keyDown(textarea, { key: 'Enter', code: 'Enter' });

  expect(mockSetComments).toHaveBeenCalledWith([
    ...props.comments,
    { username: props.currentUser, comment: 'New comment' },
  ]);
});

test('Testing if an empty comment does not trigger onSubmit()', () => {
  const mockSetComments = jest.fn();
  const props = {
    currentUser: 'Alice',
    comments: [],
    setComments: mockSetComments,
    focusOnLoad: false,
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  const textarea = getByPlaceholderText('Add a comment');

  fireEvent.change(textarea, { target: { value: '   ' } });
  fireEvent.keyDown(textarea, { key: 'Enter', code: 'Enter' });

  expect(mockSetComments).not.toHaveBeenCalled();
});

test('Testing that the CommentInput clears the comment field after submission', () => {
  const props = {
    currentUser: 'Alice',
    comments: [],
    setComments: () => {},
    focusOnLoad: false,
  };

  const { getByPlaceholderText } = render(<CommentInput {...props} />);
  const textarea = getByPlaceholderText('Add a comment');

  fireEvent.change(textarea, { target: { value: 'New comment' } });
  fireEvent.keyDown(textarea, { key: 'Enter', code: 'Enter' });

  expect(textarea.value).toBe('');
});
