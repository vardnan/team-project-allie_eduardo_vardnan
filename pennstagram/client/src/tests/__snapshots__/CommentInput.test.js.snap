// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing that the CommentInput matches snapshot 1`] = `
<div>
  <textarea
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Add a comment"
    type="text"
    value=""
  />
</div>
`;
